pandas/tests/groupby/test_function.py

```diff
@@ -286,14 +286,47 @@ def test_non_cython_api():
 
     # describe
     expected_index = pd.Index([1, 3], name="A")
-    expected_col = pd.MultiIndex(
-        levels=[["B"], ["count", "mean", "std", "min", "25%", "50%", "75%", "max"]],
-        codes=[[0] * 8, list(range(8))],
+    expected_col = pd.MultiIndex.from_product(
```

Comment:
Just to confirm: this is the result you would get on master today if you called `df.groupby("A").describe()`, rather than doing a `.mad()` on the grouper before the `describe`?

Reply:
No. Currently on master, `describe` will always drop the grouping columns, even if these are non-nuisance columns, and it does not matter if another method is called first. `describe` always drops the grouping columns. 

This behaviour is changed with this PR, so that `.describe` will now keep non-nuisance columns in the output. 

Below is sample output from running the same code on master and this PR branch. 
```
import numpy as np
from pandas import DataFrame

df = DataFrame(
    [[1, 2, "foo"], [1, np.nan, "bar"], [3, np.nan, "baz"]], columns=["A", "B", "C"]
)

print( df.groupby('A').describe() )

# Output from master
#       B
#   count mean std  min  25%  50%  75%  max
# A
# 1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

# Output from PR branch
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# A
# 1   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

`test_non_cython_api` and `test_to_latex_multiindex` are the only tests impacted by this change, but the default behaviour is changing. One of the new tests I've added explicitly enforces the new default behaviour.

Reply:
Hmm, ok... @jreback thoughts on whether to call this a bugfix or API change? We document this behavior at https://pandas.pydata.org/docs/user_guide/groupby.html#aggregation, though we don't explicitly say that the grouping column is included.

----

Ignoring API-breaking issues for a minute and just thinking about desired output, shouldn't whether `A` is included depend on `as_index`? I would expect that with `as_index=True` then `A` would be excluded, and that with `as_index=False` it would be included. Is my intuition incorrect?

Reply:
> We document this behavior at https://pandas.pydata.org/docs/user_guide/groupby.html#aggregation, though we don't explicitly say that the grouping column is included.

The example of `grouped.describe()` under this link (screenshot below) would not actually change. 

![image](https://user-images.githubusercontent.com/3813175/87822355-384e5380-c869-11ea-9c31-5f59937e0ef1.png)

This example, reimplemented below, is grouping on two `str` columns which are nuisance columns for `.describe` so they are dropped. On the new branch they will still be dropped. The new branch only breaks the API if you group on non-nuisance columns (i.e. you group on floats or ints, then call `.describe`).

```
df = pd.DataFrame({'A': ['foo', 'bar', 'foo', 'bar','foo', 'bar', 'foo', 'foo'],
                   'B': ['one', 'one', 'two', 'three','two', 'two', 'one', 'three'],
                   'C': np.random.randn(8),
                   'D': np.random.randn(8)})

df.groupby(['A', 'B'], as_index=False).describe()

# Output from new branch is unchanged.
#       C                                          ...         D
#   count      mean       std       min       25%  ...       min       25%       50%       75%       max
# 0   1.0 -0.148328       NaN -0.148328 -0.148328  ...  1.315313  1.315313  1.315313  1.315313  1.315313
# 1   1.0  0.427926       NaN  0.427926  0.427926  ...  0.547973  0.547973  0.547973  0.547973  0.547973
# 2   1.0 -0.919795       NaN -0.919795 -0.919795  ...  0.292464  0.292464  0.292464  0.292464  0.292464
# 3   2.0 -0.335763  0.121342 -0.421565 -0.378664  ...  0.480334  0.679835  0.879337  1.078839  1.278340
# 4   1.0  0.708791       NaN  0.708791  0.708791  ... -2.003574 -2.003574 -2.003574 -2.003574 -2.003574
# 5   2.0  0.554458  1.536360 -0.531913  0.011272  ... -1.134286 -0.958744 -0.783202 -0.607660 -0.432118

# [6 rows x 16 columns]
```

Reply:
> shouldn't whether `A` is included depend on `as_index`? I would expect that with `as_index=True` then `A` would be excluded, and that with `as_index=False` it would be included. Is my intuition incorrect?

Currently on master, when grouping on a non-nuisance column, `as_index` has two effects:
    `as_index=True`  --> the groups are on the index and the grouping columns are excluded from the calculation results. 
    'as_index=False` --> the groups are not on the index and the grouping column is added to the calculation results. 

On the new branch, non-nuisance grouping columns are **always** included in the calculation results of `.describe`, so `as_index` only has one effect:
    `as_index=True`  --> the groups are on the index
    `as_index=False` --> the groups are not on the index

```
import numpy as np
from pandas import DataFrame

df = DataFrame(
    [[1, 2, "foo"], [1, np.nan, "bar"], [3, np.nan, "baz"]], columns=["A", "B", "C"]
)

print( df.groupby('A', as_index=True).describe() )

# master
#       B
#   count mean std  min  25%  50%  75%  max
# A
# 1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

# new branch
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# A
# 1   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

print( df.groupby('A', as_index=False).describe() )

# master
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# 0   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 1   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

# new branch
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# 0   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 1   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

Reply:
I have updated the new test, `test_describe_results_includes_non_nuisance_columns`, so it is paramtrized over `as_index=True/False`. This will enforce that the only affect of `as_index` on `describe` is to change the index from the groups to a range index, with no other side effects.

Reply:
#28549 is related. 

I did look at having `.apply` always suppress the grouping column but this breaks vastly more tests.

Reply:
Ignore my previous comment (now deleted) if you read them. 

I've updated the PR so that the behaviour of `.describe()` is unchanged. 

```

import numpy as np
from pandas import DataFrame

df = DataFrame(
    [[1, 2, "foo"], [1, np.nan, "bar"], [3, np.nan, "baz"]], columns=["A", "B", "C"]
)

print( df.groupby('A', as_index=True).describe() )
print( df.groupby('A', as_index=False).describe() )

# Output from PR branch (same as master)
#       B
#   count mean std  min  25%  50%  75%  max
# A
# 1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# 0   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 1   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

Reply:
Following this update the `.describe` method is no longer affected by this PR. 

https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224